src/v/security/jwt.h

```diff
@@ -0,0 +1,623 @@
+}
+
+struct string_viewable_compare {
+    using is_transparent = void;
+    bool operator()(
+      string_viewable auto const& lhs, string_viewable auto const& rhs) const {
+        return char_view_cast(lhs) == char_view_cast(rhs);
+    }
+};
+
+struct string_viewable_hasher {
+    using is_transparent = void;
+    size_t operator()(string_viewable auto const& sv) const {
+        return std::hash<std::string_view>{}(char_view_cast(sv));
+    }
+};
+
+using cryptopp_bytes = ss::basic_sstring<CryptoPP::byte, uint32_t, 31, false>;
+using cryptopp_byte_view = std::basic_string_view<cryptopp_bytes::value_type>;
+
+template<typename CharT = std::string_view::value_type>
+std::basic_string_view<CharT> as_string_view(json::Value const& v) {
+    expression_in_debug_mode(
+      vassert(
+        v.IsString(), "only Strings can be converted to std::string_view"););
+    return {reinterpret_cast<CharT const*>(v.GetString()), v.GetStringLength()};
+}
+
+template<typename CharT = std::string_view::value_type>
+std::optional<std::basic_string_view<CharT>>
+string_view(json::Value const& doc, std::string_view field) {
+    auto it = doc.FindMember(field.data());
+    if (it == doc.MemberEnd() || !it->value.IsString()) {
+        return std::nullopt;
+    }
+    return as_string_view<CharT>(it->value);
+}
+
+template<string_viewable StringT = cryptopp_bytes>
+auto base64_url_decode(cryptopp_byte_view sv) {
+    CryptoPP::Base64URLDecoder decoder;
+
+    decoder.Put(sv.data(), sv.size());
+    decoder.MessageEnd();
+
+    StringT decoded;
+    CryptoPP::word64 size = decoder.MaxRetrievable();
+    if (size && size <= SIZE_MAX) {
+        decoded.resize(size);
+        decoder.Get(
+          reinterpret_cast<CryptoPP::byte*>(decoded.data()), decoded.size());
+    }
+    return decoded;
+};
+
+inline std::optional<CryptoPP::Integer>
+as_cryptopp_integer(json::Value const& v, std::string_view field) {
+    CryptoPP::Integer integer;
+    auto b64 = string_view<CryptoPP::byte>(v, field);
+    if (!b64.has_value()) {
+        return std::nullopt;
+    }
+    auto chars = base64_url_decode(b64.value());
+    return CryptoPP::Integer{chars.data(), chars.size()};
+}
+
+auto check_bytes(detail::string_viewable auto const& s) {
+    auto sv = detail::char_view_cast<char>(s);
+    return !sv.empty() && is_valid_utf8(sv) && !contains_control_character(sv);
+};
+
+template<typename Clock>
+std::optional<typename Clock::time_point>
+time_point(json::Value const& doc, std::string_view field) {
+    auto it = doc.FindMember(field.data());
+    if (it == doc.MemberEnd() || !it->value.IsInt64()) {
+        return std::nullopt;
+    }
+    return
+      typename Clock::time_point(std::chrono::seconds(it->value.GetInt64()));
+}
+
+} // namespace detail
+
+// Authorization Server Metadata as defined by
+// https://www.rfc-editor.org/rfc/rfc8414.html
+class metadata {
+public:
+    static result<metadata> make(std::string_view sv) {
+        if (!detail::check_bytes(sv)) {
```

Comment:
Why doesn't JSON parsing check for this - or why is that not enough?

Reply:
I see [kParseValidateEncodingFlag](https://rapidjson.org/namespacerapidjson.html#a81379eb4e94a0386d71d15fda882ebc9a3df81f75479da8e052393f92cbec394b), but mostly an abundance of caution.

Reply:
I'm not really sure what you're guarding against because JSON strings can be escaped, so the parsed strings can still have control characters

Reply:
My point is that this isn't enough for example: https://go.dev/play/p/OvZsawngCzT

I don't know what RapidJSON allows but even if these raw strings aren't going to be logged if you are going to log something like the kid or a value from claims it's still possible they have control characters.

Reply:
Oh, I see. I `check_bytes` after `Base64UrlDecode` but not when I've extracted a string. I'll fix that.

Reply:
Removed `check_bytes` since nothing is logged. It may be added back when extracting a principal.